File: week2/chanki/백준 1918. 후위 표기식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class B1918 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String formula = br.readLine();
        // 연산자 처리 시 괄호 -> 곱셈, 나눗셈 -> 더하기, 빼기 순으로 처리
        Deque<String> formulaWithoutBracket = removeBracket(formula);
        System.out.println(convertToPostfixNotation(formulaWithoutBracket));
    }

    private static Deque<String> removeBracket(String formula) {
        Deque<String> stack = new ArrayDeque<>();

        // 전달받은 수식을 순회하며
        for (int i = 0; i < formula.length(); i++) {
            char c = formula.charAt(i);
            // 만약 닫는 괄호라면
            if (c == ')') {
                Deque<String> elementsInBracket = new ArrayDeque<>();
                String last;
                // 여는 괄호가 나올 때까지 StringBuilder 저장
                while (!stack.isEmpty() && !(last = stack.removeLast()).equals("(")) {
                    elementsInBracket.addFirst(last);
                }
                // 거꾸로 꺼냈으므로 뒤집어서 전달
                // 연산자 처리 필요
                stack.add(convertToPostfixNotation(elementsInBracket));
            } else {
                stack.add(String.valueOf(c));
            }
        }

        return stack;
    }

    private static String convertToPostfixNotation(Deque<String> formula) {
        // 곱셈 나눗셈 먼저
        Deque<String> processedMultiplyAndDivide = new ArrayDeque<>();
        int formulaElementSize = formula.size();
        while (!formula.isEmpty()) {
            String element = formula.removeFirst();
            if (element.equals("*") || element.equals("/")) {
                processedMultiplyAndDivide.addLast(
                    processedMultiplyAndDivide.removeLast() + formula.removeFirst() + element
                );
            } else {
                processedMultiplyAndDivide.addLast(element);
            }
        }

        // 더하기 빼기 처리
        Deque<String> result = new ArrayDeque<>();
        while (!processedMultiplyAndDivide.isEmpty()) {
            String element = processedMultiplyAndDivide.removeFirst();
            if (element.equals("+") || element.equals("-")) {
                result.addLast(result.removeLast() + processedMultiplyAndDivide.removeFirst() + element);
            } else {
                result.addLast(element);
            }
        }

        StringBuilder sb = new StringBuilder();
        for (String s : result) {
            sb.append(s);
        }
        return sb.toString();
    }
}

File: week2/chanki/백준 1918. 후위 표기식.kt
import java.util.*
import kotlin.collections.ArrayDeque

// 자바와 동일하여 주석 생략
fun main() = with(Scanner(System.`in`)) {
    val formula = readlnOrNull()!!
    val deque = ArrayDeque<String>()
    for (c in formula) {
        deque.addLast(c.toString())
    }

    println(convertToPostfixFormula(removeBracket(deque)))
}

private fun removeBracket(deque: ArrayDeque<String>): ArrayDeque<String> {
    val result = ArrayDeque<String>()
    while (deque.isNotEmpty()) {
        val element = deque.removeFirst()
        when (element) {
            ")" -> {
                val bracketElements = ArrayDeque<String>()
                var bracketElement: String
                while ((result.removeLast().also { bracketElement = it }) != "(") {
                    bracketElements.addFirst(bracketElement)
                }
                result.addLast(convertToPostfixFormula(bracketElements))
            }
            else -> result.addLast(element)
        }
    }

    return result
}

private fun convertToPostfixFormula(deque: ArrayDeque<String>): String {
    val postfixMultiplyAndDivide = ArrayDeque<String>()
    while (deque.isNotEmpty()) {
        val element = deque.removeFirst()
        when {
            element == "*" || element == "/" -> {
                postfixMultiplyAndDivide.addLast(
                    postfixMultiplyAndDivide.removeLast() + deque.removeFirst() + element
                )
            }
            else -> postfixMultiplyAndDivide.addLast(element)
        }
    }

    val result = ArrayDeque<String>()
    while (postfixMultiplyAndDivide.isNotEmpty()) {
        val element = postfixMultiplyAndDivide.removeFirst()
        when {
            element == "+" || element == "-" -> {
                result.addLast(
                    result.removeLast() + postfixMultiplyAndDivide.removeFirst() + element
                )
            }
            else -> result.addLast(element)
        }
    }

    val sb = StringBuilder()
    for (s in result) {
        sb.append(s)
    }
    return sb.toString()
}